enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,549 @@
+that the customer deems as management rather than operational. It is
+important to note that not all of those components will be delivered
+as part of the OpenShift payload and some may be written by the
+customer or by vendors who are not our partners.
+
+The basic proposal is to provide a way to identify management
+workloads at runtime and to use CRI-O to run them on a user-selected
+set of CPUs, while other workloads will be prevented from running
+there. This effectively gives 3 pools of CPUs (shared, dedicated, and
+management) and means the shared CPU pool will need other CPUs to
+support burstable or best-effort workloads.
+
+We want the isolation to be in place from the initial buildout of the
+cluster, to ensure predictable behavior. Therefore, the feature must
+be enabled during installation. To enable the feature, the user will
+specify the set of CPUs to run all management workloads. When the
+management CPU set is not defined, the feature will be completely
+disabled.
+
+We generally want components to opt-in to being considered management
+workloads. Therefore, for a regular pod to be considered to contain a
+management workload it must be labeled with `io.openshift.management:
+true`.
+
+We want to treat all OpenShift components as management workloads,
+including those that run the control plane. Therefore, kubelet will be
+modified to treat all static pods as management workloads, when the
+feature is enabled. We will update operator manifests and
+implementations to label all OpenShift components not running in
+static pods with the `io.openshift.management: true` label, and add a
+CI job to require that label be present in all workloads created from
+the release payload.
+
+We need kubelet to know when the feature is enabled, but we cannot
+change the configuration schema that kubelet gets from
+upstream. Therefore we will have kubelet look for a new configuration
+file on startup, and the feature will only be enabled if that file is
+found.
+
+We want to give cluster administrators control over which workloads
+are run on the management CPUs. Therefore, only pods in namespaces
+labeled with `io.openshift.management: true` will be subject to
+special handling. Normal users cannot add a label to a namespace
+without the right RBAC permissions.
+
+We want to continue to use the scheduler for placing management
+workloads, but we cannot rely on the CPU requests of those workloads
+to accurately reflect the constrained environment in which they are
+expected to run. Instead of scaling those CPU request values, we will
+change them to request a new [extended
+resource](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/),
+"management cores." We will modify kubelet to advertise management
+cores as an extended resource when the management workload
+partitioning feature is enabled, using a value equivalent to the CPU
+resources for the entire host. This large value should allow the
```

Comment:
We're using want to advertise the full host's CPU availability just because it's a large number? In a typical case, this could easily be ~10 times the actual capacity of management cores. Are we saying this resource should essentially have unlimited availability?

If so, why do "we want to continue to use the scheduler"? What purpose does having a request for a finite amount of management core serve if the capacity is infinite?

Reply:
Oh, is it *just* for the CPU shares as described in the next paragraph? Or is the scheduler related purpose too?

Reply:
We will use the annotation on the pod to set CPU shares, but we wanted to maintain the scheduler use as well. I think that had to do with future work related to multi-node clusters, where only control plane or infrastructure nodes would advertise management cores as a resource, so management workloads would always be scheduled there. We may have other ways to achieve that, though.

Or perhaps this is another vestigial detail from the early version that talked about scaling the resources. @mrunalp or @derekwaynecarr do you remember?

Reply:
For now we want to make sure everything is scheduled. In the future, we may tune the value to something more realistic so that the scheduler distributes work in a multi-node cluster.